elasticapm/utils/disttracing.py

```diff
@@ -119,13 +122,86 @@ def merge_duplicate_headers(cls, headers, key):
         """
         # this works for all known WSGI implementations
         if isinstance(headers, list):
-            return ",".join([
-                item[1]
-                for item in headers
-                if item[0] == key
-            ])
+            return ",".join([item[1] for item in headers if item[0] == key])
         return headers.get(key)
 
+    def _parse_tracestate(self, tracestate):
+        """
+        Tracestate can contain data from any vendor, made distinct by vendor
+        keys. Vendors are comma-separated. The elastic (es) tracestate data is
+        made up of key:value pairs, separated by semicolons. It is meant to
+        be parsed into a dict.
+
+            tracestate: es=key:value;key:value...,othervendor=<opaque>
+        """
+        if not tracestate:
+            return {}
+        if "es=" not in tracestate:
+            return {}
+
+        ret = {}
+        state = re.match(r"es=([^,]*)", tracestate).group(1).split(";")
+        for keyval in state:
+            if not keyval:
+                continue
+            key, _, val = keyval.partition(":")
+            ret[key] = val
+
+        return ret
+
+    def _set_tracestate(self):
+        elastic_value = ";".join(["{}:{}".format(k, v) for k, v in compat.iteritems(self.tracestate_dict)])
+        # No character validation needed, as we validate in `add_tracestate`. Just validate length.
+        if len(elastic_value) > 256:
+            logger.debug("Modifications to TraceState would violate length limits, ignoring.")
+            raise TraceStateFormatException()
+        elastic_state = "es={}".format(elastic_value)
+        if not self.tracestate:
+            return elastic_state
+        else:
+            # Remove es=<stuff> from the tracestate, and add the new es state to the end
+            otherstate = re.sub(r"es=([^,]*),?", "", self.tracestate)
+            # No validation of `otherstate` required, since we're downstream. We only need to check `es=`
+            # since we introduced it, and that validation has already been done at this point.
+            if otherstate:
+                if otherstate[-1] == ",":
+                    return "{}{}".format(otherstate, elastic_state)
+                else:
+                    return "{},{}".format(otherstate, elastic_state)
+            else:
+                return elastic_state
```

Comment:
Should do the same, I believe

```suggestion
            return otherstate.rstrip(",") + elastic_state
```

Reply:
Almost. Still need to add a comma if `otherstate` is non-empty.